### ISSUE TITLE
Does not work with node-redis

### ISSUE THREAD

**Issue by niklasR** [aws/aws-xray-sdk-node#28]
We are trying to use the express wrapper in conjunction with a redis database.

We use [`node-redis`](https://github.com/NodeRedis/node_redis). But it does not seem to be working:
```js
const router = require('express').Router();
const redis = require('redis');
const xray = require('aws-xray-sdk');

const cache = redis.createClient({host: 'localhost'});

router.get('/test', (req, res) => {
    console.log(xray.getSegment()); // this is fine
    cache.get('test', () => {
        console.log(xray.getSegment()); // this throws
        res.json({});
    });
});
```

```
/Users/rahmen01/workspace/ibl-episodes-api/node_modules/aws-xray-sdk-core/lib/context_utils.js:21
        throw new Error(message);
        ^

Error: Failed to get the current sub/segment from the context.
    at Object.contextMissingRuntimeError [as contextMissing] (/Users/rahmen01/workspace/ibl-episodes-api/node_modules/aws-xray-sdk-core/lib/context_utils.js:21:15)
    at Object.getSegment (/Users/rahmen01/workspace/ibl-episodes-api/node_modules/aws-xray-sdk-core/lib/context_utils.js:92:45)
    at Command.cache.get [as callback] (/Users/rahmen01/workspace/ibl-episodes-api/routes/nik.js:12:26)
    at normal_reply (/Users/rahmen01/workspace/ibl-episodes-api/node_modules/redis/index.js:732:21)
    at RedisClient.return_reply (/Users/rahmen01/workspace/ibl-episodes-api/node_modules/redis/index.js:835:9)
    at JavascriptRedisParser.returnReply (/Users/rahmen01/workspace/ibl-episodes-api/node_modules/redis/index.js:193:18)
    at JavascriptRedisParser.execute (/Users/rahmen01/workspace/ibl-episodes-api/node_modules/redis-parser/lib/parser.js:574:12)
    at Socket.<anonymous> (/Users/rahmen01/workspace/ibl-episodes-api/node_modules/redis/index.js:275:27)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:250:11)
    at Socket.Readable.push (_stream_readable.js:208:10)
    at TCP.onread (net.js:607:20)
```

Same route does work fine using [`ioredis`](https://github.com/luin/ioredis):

```js
'use strict';

const router = require('express').Router();
const Redis = require('ioredis');
const xray = require('aws-xray-sdk');

const cache = new Redis();

router.get('/test', (req, res) => {
    console.log(xray.getSegment()); // this is fine
    cache.get('test', () => {
        console.log(xray.getSegment()); // this is fine
        res.json({});
    });
});

module.exports = router;
```

**Comment by niklasR** [comment]
Hello @haotianw465,

Thank you, I suspected it was something like that. Do you happen to have a hunch why `node-redis` loses the context? It does not seem to be related to the other Promise/async-await issues.

In this particular case, we use redis as cache. Depending on the outcome of the call to redis, we make additional outgoing HTTP requests to downstream APIs which we would like to capture.
In other instances, we use redis as a database, with further processing to the stored items, but the same principle applies.

It would be nice to capture the redis calls as subsegment (possibly with annotations as to whether it's an expiring cache or permanent database), but just preserving context is more important for the time being.

We'll evaluate this week if it is feasible to use `ioredis` across our estate, but right now no obvious issues come to mind (async/await support is for us definitely a higher priority).

**Comment by haotianw465** [comment]
Are you running  node v8 or later and do you promisify the client? I don't have enough time to dig into how `node-redis` implements `get` but there must be some mechanism used that let CLS loses track of the function call.

A possible workaround is to check https://docs.aws.amazon.com/xray-sdk-for-nodejs/latest/reference/ and follow the examples of `captureAsyncFunc` so you capture and annotate your `cache get` on outside instead of inside.

Another possibility of preserving context is to promisify the Redis client and patch how CLS binds context for promise by putting `AWSXRay.capturePromise();`.

**Comment by zya** [comment]
@haotianw465 Although we understand that the particular library mentioned here is not officially supported, the fact that such thing happens is an indication that the method used to persist the context (be it `cls` or otherwise) in this library is not ideal.  And this types of issues will pop up as more users start using x-ray. The issue here is not that we don't get redis related subsegment, the issue is that the code does not work when we add x-ray to the project.

**Comment by haotianw465** [comment]
We will evaluate the choice of depending on `cls` to trace context. In fact due to populate request of `async/await` we will look in to `cls-hooked` which can perserve context on `async/await` functions.

Supporting redis client is definitely in our backlog and we will make sure the context gets preserved when we developing the support. At that point we will either looking into other options of reliably tracking context or contributing to `cls` to make sure its functionality meets our customers' requirement.

**Comment by zya** [comment]
@awssandra Thanks. We do know about the manual mode, although it is possible, as you said that will require a lot of code change and we do not plan to do that atm.

@haotianw465 we did some digging around and could not find why `node-redis` does not work and loses the context. do you know the reason? Knowing that could perhaps improve handling future issues by documenting what is supported and what is not. By that, I mean similar to knowing that a language feature such as `async/await` is not supported and causes problems as opposed to documenting which `npm` packages are supported and which are not. As you know, there are millions of packages out there.

**Comment by haotianw465** [comment]
@zya  I totally understand your concerns. We will allocate some time to understand why CLS doesn't work well with `node-redis` so we can apply this knowledge to more packages. 

We will better understand the dependencies the SDK uses and improve it or even contribute to the dependencies so our customers can use more and more packages without issue. We always continue working on expanding the documented list of what is supported by testing more popular packages based on customers' feedback.

For now we recommend customers to trace libraries on our supported list. For those that are not on the list they might just work or if not you can either submit an issue so we add it to our backlog for later support or a PR is always welcome.

**Comment by awssandra** [comment]
@zya 

Looks like this is a known issue with CLS. The CLS creator has made a separate shim package to work with node-redis: https://github.com/othiym23/cls-redis

Thoughts?
Sandra

**Comment by zya** [comment]
@awssandra what we had to do to get this to work was to switch to `ioredis`. still, the underlying issue of why it does not work with `node-redis` would be more useful to investigate as it could relate to other libraries.

A complex real-world application in node would have many dependencies on many packages. Each of those would have dependencies on other packages. Keeping track of which packages are supported and not is not going to be trivial. Also, finding or implementing shims for every one of them would not be ideal.

What would be possible and helpful, is a list of supported core node patterns and functionalities that could apply to many packages.

**Comment by awssandra** [comment]
@zya

Unfortunately, CLS currently is the only viable mechanism to propagate context in the Node.js space as of now, and it is not without it's quirks. The Redis case is an example of said behavior.

From the cls-redis readme: 

> node-redis is fast. It uses a lot of smart techniques to do this that lean heavily upon Redis's architecture. One of its biggest wins is the way that it takes advantage of pipelining to batch up commands and push them to the Redis server in chunks. This works great, unless you're using CLS, which wants to provide consistent access to stored values across entire asynchronous call chains. You could use ns.bind to put all your Redis callbacks on the correct continuation chain, but that breaks down if you forget even one callback passed to client.get(). This shim's job is to take care of the bookkeeping for you. It monkeypatches the Redis driver to ensure that all the callbacks you provide are bound to the CLS namespace you provide to the shim.

The issue is that CLS has no way in knowing what async chain call matches with what namespace instance. The mechanism to track this within Node simply _doesn't exist_, which makes custom code necessary, hence the release of cls-redis which patches to manually propagate the context through each async call.

This branches out into the whole other discussions found here:
https://github.com/othiym23/node-continuation-local-storage/issues/98
https://github.com/aws/aws-xray-sdk-node/issues/12
https://github.com/othiym23/async-listener/pull/135

We cover how to use the X-Ray CLS namespace briefly in "Developing custom solutions using automatic mode" section on the core wiki: https://github.com/aws/aws-xray-sdk-node/tree/master/packages/core
This explains how to propagate the CLS namespace to develop your own solutions.
Basically, it's just wrapping in a namespace.run and setting the segment manually.

But, the fact is until Node releases it's own internal mechanism (async-hooks, which is still experimental), no solution will be 100%, and will require a shim for those other cases.
Further reading:
http://asim-malik.com/the-perils-of-node-continuation-local-storage/

Hope this helps,
Sandra

**Comment by haotianw465** [comment]
Closing in favor of https://github.com/aws/aws-xray-sdk-node/issues/60
